src/ch/ch_process.c

```diff
@@ -51,6 +53,302 @@ static virCHMonitorPtr virCHProcessConnectMonitor(virCHDriverPtr driver,
+    /*
+     * If CPU cgroup controller is not initialized here, then we need
+     * neither period nor quota settings.  And if CPUSET controller is
+     * not initialized either, then there's nothing to do anyway.
+     */
+    if (virCgroupHasController(priv->cgroup, VIR_CGROUP_CONTROLLER_CPU) ||
+        virCgroupHasController(priv->cgroup, VIR_CGROUP_CONTROLLER_CPUSET)) {
+
+        if (virDomainNumatuneGetMode(vm->def->numa, -1, &mem_mode) == 0 &&
+            mem_mode == VIR_DOMAIN_NUMATUNE_MEM_STRICT &&
+            virDomainNumatuneMaybeFormatNodeset(vm->def->numa,
+                                                priv->autoNodeset,
+                                                &mem_mask, -1) < 0)
+            goto cleanup;
+
+        if (virCgroupNewThread(priv->cgroup, nameval, id, true, &cgroup) < 0)
+            goto cleanup;
+
+        if (virCgroupHasController(priv->cgroup, VIR_CGROUP_CONTROLLER_CPUSET)) {
+            if (use_cpumask &&
+                chSetupCgroupCpusetCpus(cgroup, use_cpumask) < 0)
+                goto cleanup;
+
+            if (mem_mask && virCgroupSetCpusetMems(cgroup, mem_mask) < 0)
+                goto cleanup;
+
+        }
+
+        if ((period || quota) &&
+            chSetupCgroupVcpuBW(cgroup, period, quota) < 0)
+            goto cleanup;
+
+        /* Move the thread to the sub dir */
+        VIR_INFO("Adding pid %d to cgroup", pid);
+        if (virCgroupAddThread(cgroup, pid) < 0)
+            goto cleanup;
+
+    }
+
+    if (!affinity_cpumask)
+        affinity_cpumask = use_cpumask;
+
+    /* Setup legacy affinity. */
+    if (affinity_cpumask && virProcessSetAffinity(pid, affinity_cpumask) < 0)
+        goto cleanup;
+
+    /* Set scheduler type and priority, but not for the main thread. */
+    if (sched &&
+        nameval != VIR_CGROUP_THREAD_EMULATOR &&
+        virProcessSetScheduler(pid, sched->policy, sched->priority) < 0)
+        goto cleanup;
+
+    ret = 0;
+ cleanup:
+    if (cgroup) {
+        if (ret < 0)
+            virCgroupRemove(cgroup);
+        virCgroupFree(&cgroup);
+    }
+
+    return ret;
+}
+
+
+static int
+virCHProcessSetupEmulator(virDomainObjPtr vm)
+{
+    virCHDomainObjPrivatePtr priv = vm->privateData;
+    virCHMonitorThreadInfoPtr threads = NULL;
+    virCgroupPtr cgroup = NULL;
+    size_t nthreads;
+    int controllers;
+    int i, ret;
+
+    ret = virCHProcessSetupPid(vm, vm->pid, VIR_CGROUP_THREAD_EMULATOR,
+                               0, vm->def->cputune.emulatorpin,
+                               vm->def->cputune.emulator_period,
+                               vm->def->cputune.emulator_quota,
+                               vm->def->cputune.emulatorsched);
+    if (ret < 0)
+        return ret;
+
+    controllers = ((1 << VIR_CGROUP_CONTROLLER_CPU) |
+                   (1 << VIR_CGROUP_CONTROLLER_CPUACCT) |
+                   (1 << VIR_CGROUP_CONTROLLER_CPUSET));
+
+    if ((ret = virCgroupNewDetect(vm->pid, controllers, &cgroup)) < 0)
```

Comment:
This allocates cgroup on success but this isn't cleaned up or stored anywhere as far as I can tell.

Reply:
cgroup clean up is being handled just fine.

```
$ find /sys/fs/cgroup/ | grep machine-ch | wc -l
0
$ virsh_ch create ch_net.xml
Domain ch_net created from ch_net.xml

$ find /sys/fs/cgroup/ | grep machine-ch | wc -l
321

# Example cgroups created per VM
$ find /sys/fs/cgroup/ | grep machine-ch | cut -f 1-7 -d '/' | uniq
/sys/fs/cgroup/cpuset/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/perf_event/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/cpu,cpuacct/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/pids/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/devices/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/freezer/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/blkio/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/memory/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/net_cls,net_prio/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/systemd/machine.slice/machine-ch\x2d4153\x2dchnet.scope
/sys/fs/cgroup/unified/machine.slice/machine-ch\x2d4153\x2dchnet.scope

$ virsh_ch destroy ch_net
Domain ch_net destroyed

$ find /sys/fs/cgroup/ | grep machine-ch | wc -l
0
```
src/ch/ch_process.c:   virCHProcessStop calls chRemoveCgroup which handles cgroup clean up.

Only when libvirtd crashes or forcefully killed cgroups linger on.

Reply:
Is the`virCgroupPtr cgroup` that `virCgroupNewDetect` allocates being lost though? As far as I can tell `virCgroupFree` should be called on it since we don't use it afterwards.

Reply:
Thanks for clarifying @bryteise. I am working on handling this. I am refactoring how iothreads are handled in libvirt, that will resolve this. Will update in 1-2 days.

Reply:
I removed this custom cgroup management and adopted using virCHProcessSetupPid. This will handle cgroup creation and deletion in a standard way.